Removal of force option in snapshot create

nishith-vihar opened the pull request:
The force option does fails for snapshot create
command even though the quorum is satisfied and is redundant.
The changes removes the force option for snapshot create command
and checks if all bricks are online instead of checking for quorum
for creating a snapshot.

Fixes: #2099
Change-Id: I45d866e67052fef982a60aebe8dec069e78015bd
Signed-off-by: Nishith Vihar Sakinala <nsakinal@redhat.com>

nishith-vihar committed:
Removal of force option in snapshot create

The force option does fails for snapshot create
command even though the quorum is satisfied and is redundant.
The changes removes the force option for snapshot create command
and checks if all bricks are online instead of checking for quorum
for creating a snapshot.

Fixes: #2099
Change-Id: I45d866e67052fef982a60aebe8dec069e78015bd
Signed-off-by: Nishith Vihar Sakinala <nsakinal@redhat.com>

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

schaffung commented:
/run regression

schaffung commented:
/run regression

nishith-vihar committed:
Added flags variable for force option
flags variable is added which hold if force flag is true
or false.

Fixes: #2099
Change-Id: I2d26b8bf5066b9b528fb0e3c85e856a4268e92c9
Signed-off-by: Nishith Vihar Sakinala <nsakinal@redhat.com>

schaffung commented:
/run regression

schaffung commented:
/run regression

xhernandez commented:
I'm wondering if it wouldn't be better to accept the "force" keyword in the command and show a warning message saying that it's deprecated and will be ignored. This way we don't break backward compatibility. Otherwise existing scripts that use "force" will start failing.

schaffung commented:
> I'm wondering if it wouldn't be better to accept the "force" keyword in the command and show a warning message saying that it's deprecated and will be ignored. This way we don't break backward compatibility. Otherwise existing scripts that use "force" will start failing.

You mean the .t or ansible scripts which users might be already using?

xhernandez commented:
> You mean the .t or ansible scripts which users might be already using?

.t scripts can be easily changed. That's not a problem. The problem are all existing scripts that administrators may have created to automate things that may use the "force" keyword.

schaffung commented:
In this case, I think you're right about accepting the `force` command but displaying a deprecated warning.

schaffung reviewed (changes requested):
Accepting the `force` option in snapshot create but displaying a deprecation warning stating the reason.

nishith-vihar committed:
Deprecated message is added when force option is used
when a user uses force option along with snapshot create command
it displays that force option is now deprecated.

Fixes: 2099
Change-Id: I73caabc26dcf26cc372f2c3efde204fb803ab5de
Signed-off-by: Nishith Vihar Sakinala <nsakinal@redhat.com>

schaffung commented:
/run regression

xhernandez commented on cli/src/cli-cmd-parser.c:
```diff
@@ -4447,22 +4446,26 @@ cli_snap_create_parse(dict_t *dict, const char **words, int wordcount)
     if (strcmp(words[i], "force") == 0) {
-        flags = GF_CLI_FLAG_OP_FORCE;
-
+        ret = -1;
+        cli_err(
+            "\'force\' option is deprecated and"
+            "should not be used while creating snapshot");
+        gf_log("cli", GF_LOG_ERROR, "Invalid Syntax");
```
Wouldn't it be better to show the deprecation warning but continue normally as if `force` wasn't there ? is there any functional difference that can cause unwanted results ?

schaffung replied:
It is actually the a little different, without these code changes, the user would see that with `force`, the snapshot creation would go through irrespective of brick statuses ( i.e. even if a brick is down, the snapshot should be created , it is the quorum which matters ), but the actual implementation fails even if one brick is down. Now, the reasoning behind that was as glusterfs uses lvm to actually do these operations, there were some issues in snapshot creation when brick is down. So, it is better if we remove the force object ( for now we can go with the deprecation warning ) but whatever comparisons or checks we have in code are actually not doing anything useful with regards to snapshot and force option.

xhernandez replied:
I still don't understand. Let's assume that a user uses `force`. With this change, `force` becomes irrelevant because we check that everything is online before doing an snapshot, right ?

In that case a simple warning saying that `force` is deprecated and the snapshot will only succeed if all bricks are online should be sufficient, and the operation will proceed. We don't need to fail the command simply because `force` is there. If a brick is down, the command will fail anyway, and that's fine, because we have warned about it.

So, if we keep the `force` option for backward compatibility, we must not fail if it's present saying that it's an invalid syntax (otherwise it's not backward compatible. It's the same as removing the option). A deprecation warning and proceeding as if it was not present seems the most optimal solution to me.

nishith-vihar committed:
Snapshot is created with force option if all bricks are online
Deprecated message is added for force option so that
when force option is used it displays deprecated message
and the snapshot is created if all the bricks are online.
A to-do is added for removing the force option in the
further releases.

Fixes: #2099
Change-Id: I0d950c7278ea4d595405d0828b5ad049afb6c27f
Signed-off-by: Nishith Vihar Sakinala <nsakinal@redhat.com>

schaffung commented:
/run regression

xhernandez commented on cli/src/cli-cmd-parser.c:
```diff
@@ -4448,22 +4447,24 @@ cli_snap_create_parse(dict_t *dict, const char **words, int wordcount)
     if (strcmp(words[i], "force") == 0) {
-        flags = GF_CLI_FLAG_OP_FORCE;
-
+        cli_out(
+            "Warning: \'force\' option is deprecated and "
+            "should not be used in the future while "
+            "creating snapshot");
```
I would add more information here or a link where more information can be found (now `force` is a noop and snapshot is only allowed if all bricks are online).

I would also add a `TODO` to remember to completely remove the `force` keyword after a couple of releases.

nishith-vihar replied:
Added a Todo and some info regarding all bricks beings online for command to execute.

xhernandez reviewed (commented):
Could you also update the commit message to reflect the real change in this patch ?

schaffung commented:
/run regression
